enumerables.rb

```diff
@@ -0,0 +1,168 @@
+module Enumerable
+  def my_each
+    return enum_for(__callee__) unless block_given?
+
+    size.times { |i| yield to_a[i] }
+  end
+
+  def my_each_with_index
+    return enum_for(__callee__) unless block_given?
+
+    size.times { |i| yield(to_a[i], i) }
+  end
+
+  def my_select
+    return enum_for(__callee__) unless block_given?
+
+    res = []
+    my_each { |i| res << i if yield(i) }
+    res
+  end
+
+  def my_all?
+    return enum_for(__callee__) unless block_given?
+
+    my_each { |i| return false unless yield(i) }
+    true
+  end
+
+  def my_any?
+    return enum_for(__callee__) unless block_given?
+
+    my_each { |i| return true if yield(i) }
+    false
+  end
+
+  def my_none?
+    return enum_for(__callee__) unless block_given?
+
+    my_each { |i| return false if yield(i) }
+    true
+  end
+
+  def my_count(cond = nil)
+    c = 0
+    if block_given?
+
+      my_each { |i| c += 1 if yield(i) }
+    else
+      my_each { |i| c += 1 if i == cond }
+
+    end
+    c
+  end
+
+  def my_map(&oper)
+    res = []
+    if block_given?
+      my_each { |i| res << yield(i) }
+    else
+      my_each { |i| res << oper.call(i) }
+    end
+    res
+  end
```

Comment:
```rb
2.6.5 :082 > array
 => [4, 1, 4, 4, 5, 3, 0, 4, 2, 8, 3, 4, 8, 0, 2, 0, 4, 2, 7, 0, 5, 4, 3, 3, 8, 8, 5, 3, 8, 8, 5, 4, 7, 6, 2, 1, 4, 1, 8, 1, 4, 1, 1, 5, 6, 8, 8, 5, 6, 0, 6, 8, 7, 6, 2, 0, 5, 5, 6, 6, 4, 5, 4, 2, 5, 0, 0, 6, 2, 5, 4, 4, 4, 7, 1, 0, 3, 8, 1, 3, 7, 8, 8, 4, 7, 5, 6, 0, 5, 2, 0, 5, 6, 3, 5, 2, 8, 7, 2, 1] 
2.6.5 :083 > array.my_map
Traceback (most recent call last):
        9: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        7: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        6: from (irb):83
        5: from enumerables.rb:60:in `my_map'
        4: from enumerables.rb:5:in `my_each'
        3: from enumerables.rb:5:in `times'
        2: from enumerables.rb:5:in `block in my_each'
        1: from enumerables.rb:60:in `block in my_map'
NoMethodError (undefined method `call' for nil:NilClass)
2.6.5 :084 > array.map
 => #<Enumerator: [4, 1, 4, 4, 5, 3, 0, 4, 2, 8, 3, 4, 8, 0, 2, 0, 4, 2, 7, 0, 5, 4, 3, 3, 8, 8, 5, 3, 8, 8, 5, 4, 7, 6, 2, 1, 4, 1, 8, 1, 4, 1, 1, 5, 6, 8, 8, 5, 6, 0, 6, 8, 7, 6, 2, 0, 5, 5, 6, 6, 4, 5, 4, 2, 5, 0, 0, 6, 2, 5, 4, 4, 4, 7, 1, 0, 3, 8, 1, 3, 7, 8, 8, 4, 7, 5, 6, 0, 5, 2, 0, 5, 6, 3, 5, 2, 8, 7, 2, 1]:map> 
2.6.5 :085 > 
```

Reply:
```rb
2.6.5 :087 > my_proc = proc {|num| num > 10 }
 => #<Proc:0x00000000026d5388@(irb):87> 
2.6.5 :088 > array
 => [4, 1, 4, 4, 5, 3, 0, 4, 2, 8, 3, 4, 8, 0, 2, 0, 4, 2, 7, 0, 5, 4, 3, 3, 8, 8, 5, 3, 8, 8, 5, 4, 7, 6, 2, 1, 4, 1, 8, 1, 4, 1, 1, 5, 6, 8, 8, 5, 6, 0, 6, 8, 7, 6, 2, 0, 5, 5, 6, 6, 4, 5, 4, 2, 5, 0, 0, 6, 2, 5, 4, 4, 4, 7, 1, 0, 3, 8, 1, 3, 7, 8, 8, 4, 7, 5, 6, 0, 5, 2, 0, 5, 6, 3, 5, 2, 8, 7, 2, 1] 
2.6.5 :089 > array.map(my_proc)
Traceback (most recent call last):
        5: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        3: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):89
        1: from (irb):89:in `map'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :090 > array.map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :091 > array.my_map(&my_proc)
 => [false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false] 
2.6.5 :092 > array.map(&my_proc) == array.my_map(&my_proc)
 => true 
2.6.5 :093 > 
2.6.5 :094 > array.my_map(my_proc) {|num| num < 10 }
Traceback (most recent call last):
        5: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        4: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        3: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):94
        1: from enumerables.rb:55:in `my_map'
ArgumentError (wrong number of arguments (given 1, expected 0))
2.6.5 :095 > 
```

Reply:
Fixed when we updated the my_each method